p2p/ipld/nmt_wrapper.go

```diff
@@ -0,0 +1,91 @@
+package ipld
+
+import (
+	"crypto/sha256"
+
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/nmt"
+	"github.com/lazyledger/nmt/namespace"
+	"github.com/lazyledger/rsmt2d"
+)
+
+// Fulfills the rsmt2d.Tree interface and rsmt2d.TreeConstructorFn function
+var _ rsmt2d.TreeConstructorFn = ErasuredNamespacedMerkleTree{}.Constructor
+var _ rsmt2d.Tree = &ErasuredNamespacedMerkleTree{}
+
+// ErasuredNamespacedMerkleTree wraps NamespaceMerkleTree to conform to the
+// rsmt2d.Tree interface while catering specifically to erasure data. For the
+// first half of the tree, it uses the first DefaultNamespaceIDLen number of
+// bytes of the data pushed to determine the namespace. For the second half, it
+// uses the parity namespace ID
+type ErasuredNamespacedMerkleTree struct {
+	squareSize uint64
+	pushCount  uint64
+	options    []nmt.Option
+	tree       *nmt.NamespacedMerkleTree
+}
+
+// NewErasuredNamespacedMerkleTree issues a new ErasuredNamespacedMerkleTree
+func NewErasuredNamespacedMerkleTree(squareSize uint64, setters ...nmt.Option) ErasuredNamespacedMerkleTree {
+	return ErasuredNamespacedMerkleTree{squareSize: squareSize, options: setters}
+}
+
+// Constructor acts as the rsmt2d.TreeConstructorFn for
+// ErasuredNamespacedMerkleTree
+func (w ErasuredNamespacedMerkleTree) Constructor() rsmt2d.Tree {
+	w.tree = nmt.New(sha256.New(), w.options...)
+	return &w
+}
+
+// Push adds the provided data to the underlying NamespaceMerkleTree, and
+// automatically uses the first DefaultNamespaceIDLen number of bytes as the
+// namespace unless the data pushed to the second half of the tree. Fulfills the
+// rsmt.Tree interface. NOTE: panics if there's an error pushing to underlying
+// NamespaceMerkleTree or if the tree size is exceeded
+func (w *ErasuredNamespacedMerkleTree) Push(data []byte) {
+	// determine the namespace based on where in the tree we're pushing
+	nsID := make(namespace.ID, types.NamespaceSize)
+
+	switch {
+	// panic if the tree size is exceeded
+	case w.pushCount > 2*w.squareSize:
```

Comment:
Is there a test for this? Doesn't look like this case is covered.

```suggestion
	case w.pushCount >= 2*w.squareSize:
```

Reply:
Good catch! Tja am Yeah, that should probably be covered by test-case. IMO the caller should simply provide the parity bytes. But that is probably a matter of taste (adding the parity namespaces has to happen somewhere). Nur let's stick with the chosen approach and add a test-case instead.

Reply:
this panic is covered in the first test case of `TestErasureNamespacedMerkleTreePanics`, but it needed the `>=`! thanks